packages/fxa-auth-server/lib/payments/paypal-processor.ts

```diff
@@ -0,0 +1,335 @@
+      if (successTransactions.length > 1) {
+        this.log.error('multipleCompletedTransactions', {
+          customer,
+          invoiceId: invoice.id,
+          transactionCount: successTransactions.length,
+          excessTransactions: successTransactions
+            .slice(1)
+            .map((t) => t.transactionId),
+        });
+        reportSentryError(
+          new Error('Multiple completed payments for invoice: ' + invoice.id)
+        );
+      }
+      const transaction = successTransactions[0];
+      await Promise.all([
+        this.stripeHelper.updateInvoiceWithPaypalTransactionId(
+          invoice,
+          transaction.transactionId
+        ),
+        this.stripeHelper.payInvoiceOutOfBand(invoice),
+      ]);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Handles the condition where a transaction for an invoice is pending.
+   *
+   * This occurs when a previous payment attempt has not completed. This
+   * method checks to see if there's a pending transaction for the invoice.
+   * Returns `true` if this invoice has valid pending transactions to wait
+   * for, `false` otherwise.
+   *
+   * @param invoice
+   * @param transactions
+   */
+  private async handlePendingTransaction(
+    invoice: Stripe.Invoice,
+    transactions: TransactionSearchResult[]
+  ): Promise<boolean> {
+    const customer = invoice.customer;
+    const inGracePeriod = this.inGracePeriod(invoice);
+    const outstandingTransactions = transactions.filter((t) =>
+      ['Pending', 'In-Progress'].includes(t.status)
+    );
+    if (outstandingTransactions.length) {
+      if (outstandingTransactions.length > 1) {
+        this.log.error('multiplePendingTransactions', {
+          customer,
+          invoiceId: invoice.id,
+        });
+        reportSentryError(
+          new Error('Multiple pending payments for invoice: ' + invoice.id)
+        );
+      }
+      if (inGracePeriod) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Make a payment attempt on an invoice.
+   *
+   * @param invoice
+   */
+  /**
+   * Make a payment attempt on an invoice.
+   *
+   * @param invoice
+   */
+  private async makePaymentAttempt(invoice: Stripe.Invoice) {
+    const customer = invoice.customer as Stripe.Customer;
+    if (invoice.amount_due === 0) {
+      return this.paypalHelper.processZeroInvoice(invoice);
+    }
+
+    try {
+      await this.paypalHelper.processInvoice({
+        customer,
+        invoice,
+        batchProcessing: true,
+      });
+    } catch (err) {
+      if (err instanceof PayPalClientError) {
+        if (err.errorCode === PAYPAL_BILLING_AGREEMENT_INVALID) {
+          const uid = customer.metadata.userid;
+          await Promise.all([
+            this.stripeHelper.removeCustomerPaypalAgreement(customer),
+            updatePayPalBA(
+              uid,
+              this.stripeHelper.getCustomerPaypalAgreement(customer) as string,
+              'Cancelled',
```

Comment:
So I learned from a l10n PR that apparently 'Cancelled' is the Canadian spelling of 'Canceled'?!

Anyways maybe we can do something like 
```
export const PayPalBillingAgreementStatuses = {
  PENDING: 'Pending',
  ACTIVE: 'Active',
  SUSPENDed: 'Suspended',
  CANCELLed: 'Cancelled',
  EXPIRED: 'Expired',
} as const;
type PayPalBillingAgreementStatuses = typeof PayPalBillingAgreementStatuses;
export type PayPalBillingAgreementStatusType = PayPalBillingAgreementStatuses[keyof PayPalBillingAgreementStatuses];
```

then we can use `PayPalBillingAgreementStatuses.CANCELLED` instead of a literal here.

Reply:
I wonder if this means PayPal is originally Canadian. I should note that the `updatePayPalBA` function does have a strong type for the string, TS will fail to compile if you typo that string and VS Code gives proper completions there.